Makefile

```diff
@@ -1,199 +1,342 @@
-  TEMP_FILE := $$(FULL_OUTPUT_DIR)/$(1).tmp.s
-
-  ifeq ($(findstring _NoN,$(1)),)
-    CUR_NoN := 0
-  else
-    CUR_NoN := 1
-  endif
-  
-  ifneq ($(findstring _XBUS,$(1)),)
-    METHOD := $(METHOD_XBUS)
-  else
-    METHOD := $(METHOD_FIFO)
-  endif
-
-  NAME := $(NAME_$(1))
-  ID := $$(ID_$$(CUR_UCODE_WITHOUT_SUFFIX))
-  TYPE := $$(TYPE_$$(CUR_UCODE))
-  CODE_MD5 := $$(MD5_CODE_$(1))
-  DATA_MD5 := $$(MD5_DATA_$(1))
-endef
-
-# Sets up the microcode make rules
-define ucode_rule
-  $(eval $(call set_vars,$(1)))
-
-  $(CODE_FILE) $(DATA_FILE) $(SYM_FILE) $(SYM2_FILE) $(TEMP_FILE): ./f3dex2.s ./rsp/* $(FULL_OUTPUT_DIR)
-	@printf "$(INFO)Building microcode: $(FULL_UCODE)$(NO_COL)\n"
-	@$(ARMIPS) -strequ DATA_FILE $(DATA_FILE) -strequ CODE_FILE $(CODE_FILE) -strequ NAME "$(NAME)" -equ UCODE_TYPE $(TYPE) -equ UCODE_METHOD $(METHOD) -equ UCODE_ID $(ID) -equ NoN $(CUR_NoN) f3dex2.s -sym2 $(SYM_FILE) -temp $(TEMP_FILE)
-  ifeq ($(CODE_MD5),)
-	@printf "  $(WARNING)Nothing to compare $(1) to!$(NO_COL)\n"
-  else
-	@(printf "$(CODE_MD5) *$(CODE_FILE)" | md5sum --status -c -) && printf "  $(SUCCESS)$(1) code matches$(NO_COL)\n" || printf "  $(FAILURE)$(1) code differs$(NO_COL)\n"
-	@(printf "$(DATA_MD5) *$(DATA_FILE)" | md5sum --status -c -) && printf "  $(SUCCESS)$(1) data matches$(NO_COL)\n" || printf "  $(FAILURE)$(1) data differs$(NO_COL)\n"
-
-  endif
-
-  $(FULL_OUTPUT_DIR): | $(OUTPUT_DIR)
-	@printf "$(INFO)Creating directory: $(FULL_OUTPUT_DIR)$(NO_COL)\n"
-ifeq ($(OS),Windows_NT)
-	@mkdir $(FULL_OUTPUT_DIR)
-else
-	@mkdir -p $(FULL_OUTPUT_DIR)
-endif
-
-  FULL_OUTPUT_DIRS += $(FULL_OUTPUT_DIR)
-  CODE_FILES += $(CODE_FILE)
-endef
-
-ifeq ($(NoN),1)
-  SUFFIX := _NoN
-else
-  SUFFIX := 
-endif
-INPUT_UCODE := $(UCODE)$(SUFFIX)_$(VERSION)
-FULL_OUTPUT_DIR := $(OUTPUT_DIR)/$(INPUT_UCODE)
-ifeq ($(OS),Windows_NT)
-  FULL_OUTPUT_DIR := $(subst /,\,$(FULL_OUTPUT_DIR))
-endif
-CODE_FILE := $(FULL_OUTPUT_DIR)/$(INPUT_UCODE).code
-DATA_FILE := $(FULL_OUTPUT_DIR)/$(INPUT_UCODE).data
-
-
-default: $(CODE_FILE) $(DATA_FILE)
-
-$(foreach ucode,$(UCODES),$(eval $(call ucode_rule,$(ucode))))
-
-all: $(CODE_FILES)
-
-$(OUTPUT_DIR):
-	@printf "$(INFO)Creating output directory$(NO_COL)\n"
-ifeq ($(OS),Windows_NT)
-	@mkdir $@
-else
-	@mkdir -p $@
-endif
-
-clean:
-	@printf "$(WARNING)Deleting all built microcode files$(NO_COL)\n"
-	@rm -rf $(FULL_OUTPUT_DIRS)
-
-.PHONY: default check all clean
+MAKEFLAGS += --no-builtin-rules
+MAKEFLAGS += --no-builtin-variables
+.SUFFIXES:
+
+default:
+	@echo 'How to use this Makefile'
+	@echo 'Method 1) Select a microcode, just run e.g. `make F3DZEX_NoN_2.06H`'
+	@echo 'Method 2) `make ok`, builds all ucodes which have MD5 sums, to check if :OK:'
+	@echo 'Method 3) `make all`, builds all ucodes in database'
+	@echo 'Method 4) Custom microcode build via custom.mk file. Create a file custom.mk'
+	@echo 'with contents like this:'
+	@echo 'NAME := your_custom_ucode_name'
+	@echo 'GAMES := Your Romhack Name'
```

Comment:
It seems like GAMES may be better named something along the lines of description, since some ucodes have too many games to list. Also, some ucodes don't show up in a game so a description could populate this field instead explaining what the ucode is, like the no-NoN variants of F3DZEX2.

Reply:
Good idea, changed.